sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -106,7 +106,27 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   final def execute(): RDD[InternalRow] = {
     RDDOperationScope.withScope(sparkContext, nodeName, false, true) {
       prepare()
-      doExecute()
+
+      if (sqlContext.conf.wholeStageEnabled && supportCodeGen
```

Comment:
It might be nice to actually make this a first class concept in the planner.  Perhaps a phase that collects valid nodes and replaces them with an explicit operator called `WholeStageCodeGen` that holds all the operators that are getting.  I'm not sure how hard that would be, but it would make debugging this a lot easier if it was visible from explain.